linker/shared/src/main/scala/org/scalajs/linker/backend/javascript/SourceMapWriter.scala

```diff
@@ -171,52 +193,174 @@ object SourceMapWriter {
-    private val elements = new ArrayBuffer[FragmentElement]
+  final class FragmentBuilder(index: Index) extends Builder(index) {
+    private val data = new ByteArrayWriter()
+
+    private var lastColumnInGenerated = 0
+    private var lastSource: SourceFile = null
+    private var lastSourceIndex = 0
+    private var lastLine: Int = 0
+    private var lastColumn: Int = 0
+    private var lastNameIndex: Int = 0
 
     protected def doWriteNewLine(): Unit =
-      elements += FragmentElement.NewLine
+      data.write(FragNewLine)
+
+    protected def doWriteSegment(columnInGenerated: Int, source: SourceFile,
+        line: Int, column: Int, name: String): Unit = {
+      val MaxSegmentLength = 1 + 5 * 5 // segment type + max 5 rawVLQ of max 5 bytes each
+      val buffer = data.unsafeStartDirectWrite(maxBytes = MaxSegmentLength)
+      var offset = data.currentSize
+
+      // Write segment type
+      buffer(offset) = {
+        if (source == null) FragColOnly
+        else if (name == null) FragColAndPos
+        else FragColPosName
+      }
+      offset += 1
+
+      offset = writeRawVLQ(buffer, offset, columnInGenerated-lastColumnInGenerated)
+      lastColumnInGenerated = columnInGenerated
+
+      if (source != null) {
+        if (source eq lastSource) { // highly likely
+          buffer(offset) = 0
+          offset += 1
+        } else {
+          val sourceIndex = index.sourceToIndex(source)
+          offset = writeRawVLQ(buffer, offset, sourceIndex-lastSourceIndex)
+          lastSource = source
+          lastSourceIndex = sourceIndex
+        }
+
+        // Line field
+        offset = writeRawVLQ(buffer, offset, line - lastLine)
+        lastLine = line
 
-    protected def doWriteSegment(columnInGenerated: Int, pos: Position, name: String): Unit =
-      elements += FragmentElement.Segment(columnInGenerated, pos, name)
+        // Column field
+        offset = writeRawVLQ(buffer, offset, column - lastColumn)
+        lastColumn = column
 
-    protected def doComplete(): Unit = {
-      if (elements.nonEmpty && elements.last != FragmentElement.NewLine)
-        throw new IllegalStateException("Trying to complete a fragment in the middle of a line")
+        // Name field
+        if (name != null) {
+          val nameIndex = index.nameToIndex(name)
+          offset = writeRawVLQ(buffer, offset, nameIndex-lastNameIndex)
+          lastNameIndex = nameIndex
+        }
+      }
+      data.unsafeEndDirectWrite(offset)
+    }
+
+    protected def doComplete(): Unit = ()
+
+    private def writeRawVLQ(buffer: Array[Byte], offset0: Int, value0: Int): Int = {
+      // See comment in writeBase64VLQ
+      val signExtended = value0 >> 31
+      var value = (((value0 ^ signExtended) - signExtended) << 1) | (signExtended & 1)
+
+      var offset = offset0
+
+      while ({
+        if ((value & ~127) != 0)
+          buffer(offset) = ((value & 127) | 128).toByte
+        else
+          buffer(offset) = (value & 127).toByte
```

Comment:
Consider using hexadecimal notation when we use numbers for their bit patterns:
```suggestion
        if ((value & ~0x7f) != 0)
          buffer(offset) = ((value & 0x7f) | 0x80).toByte
        else
          buffer(offset) = (value & 0x7f).toByte
```